OpenStreetMap data parser

dbrgn opened the pull request:
This PR adds parsing for OpenStreetMap data source. It can load a charging station from JSON and convert it to a `ChargeLocation` instance.

Supported values:

- Name (tag: `name`)
- Description (tag: `description`)
- Location (lat/lon)
- Network (tag: `network`)
- Operator (tag: `operator`)
- Chargepoints, including type and output power
- URL and Edit URL
- Cost (`fee` and `parking:fee`)
- Some opening hours (only `opening_hours=24/7`)
- License

Tests are provided, but the class is not yet integrated into the app itself. However, this should pave the way for a new OSM data source (#97).

Notes and open issues:

- Based on #159, merge that one first and then I can rebase (and mark this as non-draft)
- Address is not yet known and defaults to empty values (maybe this should be nullable?)
- What does `barrierFree` entail? Probably it should be named `accessible`, but that's still unclear. Parking lot for people with disabilities? Reachability with a wheelchair? See https://wiki.openstreetmap.org/wiki/Disabilities for different meanings.
- Right now the chargepoint type is represented using a human-readable string like "CCS (either Type 1 or Type 2)". The possible values are assigned to constants in the `Chargepoint` companion object. However, that's really not a clean solution. Instead, there should probably be an enum that lists all known chargepoint types and that has a clearly defined representation for serialization.

johan12345 commented:
Thank you! 🎉 

Some comments about the open issues:

> Address is not yet known and defaults to empty values (maybe this should be nullable?)
Yeah, I think we can make the address nullable (if it can't easily be determined from the OSM data).

> What does barrierFree entail? Probably it should be named accessible, but that's still unclear. Parking lot for people with disabilities? Reachability with a wheelchair? See https://wiki.openstreetmap.org/wiki/Disabilities for different meanings.

No, sorry that this was not yet documented: `barrierFree` means that the charger offers the possibility of direct/ad-hoc payment (with a credit/debit card or cash etc.) without the need to sign up e.g. for an RFID card or app from a service provider. Most chargepoint operators implement this using a QR code on the charger that leads to a payment website. But it could also be a credit card terminal or a coin slot directly on the charger, or payment in a shop. Free chargers can probably also be considered `barrierFree` if activating them does not require a registration.

GoingElectric labels this attribute "Barrierefreies Laden" or "ohne Vertrag/Registrierung nutzbar" in their UI and calls it `barrierFree` in the API. 

> Right now the chargepoint type is represented using a human-readable string like "CCS (either Type 1 or Type 2)". The possible values are assigned to constants in the Chargepoint companion object. However, that's really not a clean solution. Instead, there should probably be an enum that lists all known chargepoint types and that has a clearly defined representation for serialization.

The initial idea behind using a string instead of an enum there was that the list of plug/socket types used all over the world would become very long, especially when including all the different kinds of household sockets. So, in case a plug provided by the data source is not supported by EVMap, we can still show it as a free text field. But I also agree it would be cleaner with an enum... Maybe an enum field with an option `OTHER` combined with a string field with further information might be a solution?

johan12345 commented on app/src/main/java/net/vonforst/evmap/api/openstreetmap/OpenStreetMapModel.kt:
```diff
@@ -0,0 +1,191 @@
+        // Note: The current {@link OpeningHours} format is not flexible enough to handle
+        // all rules that OSM can represent and might need to be updated.
+        // This library could help:
+        //
```
link is missing here? :)

dbrgn replied:
Good catch 🙂

johan12345 commented on app/src/main/java/net/vonforst/evmap/api/openstreetmap/OpenStreetMapModel.kt:
```diff
@@ -0,0 +1,191 @@
+         * instead of "kW", those can be treated as equivalent.
+         *
+         * Sometimes people also mapped plain numbers (e.g. 7000, I assume that's 7 kW),
+         * ranges (5,5 - 11 kW, huh?) or even power (32 A), which is wrong. If we cannot parse,
```
"`or even power (32 A)`" -> "`or even amperage (32 A)`"

dbrgn replied:
Ah, yeah, that's a mis-translation in my head 🙈

dbrgn commented:
> No, sorry that this was not yet documented: `barrierFree` in this context means that the charger offers the possibility of direct/ad-hoc payment (with a credit/debit card or cash etc.) without the need to sign up e.g. for an RFID card or app from a service provider.

Ahhh, that's [`authentication:none=yes`](https://wiki.openstreetmap.org/wiki/Key:authentication) in OSM 🙂 I'll integrate that into the branch.

dbrgn commented:
I did some initial fixes, will finish them later today.

TODO:

- [ ] Make address nullable
- [ ] Finish documentation of ChargeLocation fields

The chargepoint type can be handled in a separate PR.

dbrgn commented on app/src/main/java/net/vonforst/evmap/model/ChargersModel.kt:
```diff
@@ -24,6 +24,32 @@ import kotlin.math.floor
+ * @param operator The operator of this charge location (Charge Point Operator, CPO)
+ * @param generalInformation TODO
+ * @param amenities TODO
+ * @param locationDescription TODO
```
@johan12345 Do you have a good description for these three fields? I was a bit unsure about them.

johan12345 replied:
GoingElectric labels these fields as "Allgemeine Information (Informationen, für die es keinen eigenen Punkt gibt)", "Ladeweile" and "Beschreibung der Lage / Position". So maybe something like:
```
* @param generalInformation General information about this charging site that does not fit anywhere else
* @param amenities Description of amenities available at or near the charging site (toilets, food, accommodation, landmarks, etc.)
* @param locationDescription Directions on how to find the charger (e.g. "In the parking garage on level 5")
```

dbrgn replied:
Sounds good!

johan12345 commented:
@dbrgn I'm now merging #157, which also adds two new fields `timeRetrieved` and `isDetailed` to the `ChargeLocation` class. These fields will be useful for offline data storage/caching, but will probably cause some conflicts when rebasing this PR. `isDetailed` can probably always be set to `true` for OSM data as we have no compact representation of the OSM data that includes fewer details.

dbrgn commented:
Thanks for the helpful rebase notes 🙂

Should `timeRetrieved` be set to the time the OSM data extract was generated? Or to the time it was parsed? The former probably makes the most sense.

We have access to three additional properties that might be interesting: The timestamp of the last update in OSM, the version (monotonically increasing integer) and the username that last edited this node. Do you think we should integrate these somehow? The information could be integrated into the general description, e.g. "information last updated in OpenStreetMap by <username> <n> days ago", but this is hard to localize. Additionally, just because an entry is old doesn't mean it's wrong. (Note: This could also be handled in a future PR.)

dbrgn commented:
Rebased against master and cherry picked some commits from other open PRs. Can be rebased again once #159 and #161 are merged.

I _think_ all previously open issues are now addressed.

johan12345 commented:
Thanks! 😊 I'll try to take a look at this later today, but it might also take until tomorrow.

johan12345 commented:
Yep, I agree, the open issues are addressed 🙂 Can you rebase again as #159 and #161 are merged now?

> Should timeRetrieved be set to the time the OSM data extract was generated? Or to the time it was parsed? The former probably makes the most sense.

Good question. I was thinking to use this for reminding the user to re-download their offline data - for GoingElectric this would need to be already after 24h due to their API usage policy, but for OSM data there is probably no such policy - so we could remind the user e.g. after 30 days.
So if we update the OSM extract every day or so, it should not make a huge difference whether we use the extraction time or the parsing time. But yes, if we show that value to the user and a change was made on OSM shortly after the extract was generated it might be confusing, so using the extraction time might be the best solution.

> The timestamp of the last update in OSM, the version (monotonically increasing integer) and the username that last edited this node. Do you think we should integrate these somehow? The information could be integrated into the general description, e.g. "information last updated in OpenStreetMap by days ago", but this is hard to localize. Additionally, just because an entry is old doesn't mean it's wrong. (Note: This could also be handled in a future PR.)

Yeah, this would be interesting to add in a future PR. Not sure if we need the version number, but the last edited date and username would definitely be nice. We could e.g. show this at the bottom of the charger details screen where we already have the licensing info.

dbrgn commented:
@johan12345 great, branch is now rebased and marked as ready!
